op-supervisor/supervisor/backend/backend.go

```diff
@@ -60,61 +68,115 @@ func NewSupervisorBackend(ctx context.Context, logger log.Logger, m Metrics, cfg
+	chainMetrics := make(map[types.ChainID]*chainMetrics, len(chains))
 
 	// create the supervisor backend
 	super := &SupervisorBackend{
 		logger:          logger,
 		m:               m,
 		dataDir:         cfg.Datadir,
 		depSet:          depSet,
+		chainDBs:        chainsDBs,
 		chainProcessors: chainProcessors,
-		db:              chainsDB,
+		chainMetrics:    chainMetrics,
+		// For testing we can avoid running the processors.
+		synchronousProcessors: cfg.SynchronousProcessors,
+	}
+
+	// Initialize the resources of the supervisor backend.
+	// Stop the supervisor if any of the resources fails to be initialized.
+	if err := super.initResources(ctx, cfg); err != nil {
+		err = fmt.Errorf("failed to init resources: %w", err)
+		return nil, errors.Join(err, super.Stop(ctx))
 	}
 
-	// from the RPC strings, have the supervisor backend create a chain monitor
-	// don't start the monitor yet, as we will start all monitors at once when Start is called
+	return super, nil
+}
+
+// initResources initializes all the resources, such as DBs and processors for chains.
+// An error may returned, without closing the thus-far initialized resources.
+// Upon error the caller should call Stop() on the supervisor backend to clean up and release resources.
+func (su *SupervisorBackend) initResources(ctx context.Context, cfg *config.Config) error {
+	chains := su.depSet.Chains()
+
+	// for each chain known to the dependency set, create the necessary DB resources
+	for _, chainID := range chains {
+		if err := su.openChainDBs(chainID); err != nil {
+			return fmt.Errorf("failed to open chain %s: %w", chainID, err)
+		}
+	}
+
+	// for each chain initialize a chain processor service
+	for _, chainID := range chains {
+		logProcessor := processors.NewLogProcessor(chainID, su.chainDBs)
+		chainProcessor := processors.NewChainProcessor(su.logger, chainID, logProcessor, su.chainDBs)
+		su.chainProcessors[chainID] = chainProcessor
+	}
+
+	// the config has some RPC connections to attach to the chain-processors
 	for _, rpc := range cfg.L2RPCs {
-		err := super.addFromRPC(ctx, logger, rpc, false)
+		err := su.attachRPC(ctx, rpc)
 		if err != nil {
-			return nil, fmt.Errorf("failed to add chain monitor for rpc %v: %w", rpc, err)
+			return fmt.Errorf("failed to add chain monitor for rpc %v: %w", rpc, err)
 		}
 	}
-	return super, nil
+	return nil
 }
 
-// addFromRPC adds a chain monitor to the supervisor backend from an rpc endpoint
-// it does not expect to be called after the backend has been started
-// it will start the monitor if shouldStart is true
-func (su *SupervisorBackend) addFromRPC(ctx context.Context, logger log.Logger, rpc string, _ bool) error {
-	// create the rpc client, which yields the chain id
-	rpcClient, chainID, err := clientForL2(ctx, logger, rpc)
+// openChainDBs initializes all the DB resources of a specific chain.
+// It is a sub-task of initResources.
+func (su *SupervisorBackend) openChainDBs(chainID types.ChainID) error {
```

Comment:
When I mentioned that there's a database per chain, there's actually multiple:
* LogsDB, where the raw log data is stored
* Local Derived From, where L1 derivation information is stored
* Cross Derived From, where more L1 derivation information is held